### PR TITLE
attempt to fix #663

### LINKED ISSUE #663: Crunching space around alterations mid-line
Much like alterations which lead a line, it should be possible for alterations which lead a syllable mid-line to invade the space of the previous syllable so long as a few conditions are met:

1. The alteration extends out to the left of the lyrics of its own syllable (i.e. the syllable has a positive `begindifference`)
2. The previous syllable had lyrics which extended beyond the end of its notes (i.e. `previousendifference` is positive)
3. The space between the notes of the previous syllable and the alteration is at least `spacebeforealteration`.

### PR DESCRIPTION
**Pull request by henryso**
This is a rebasing of #686 onto release-4.0.

Note: when I rebased the change to release-4.0, it closed the original pull request (#686).  There is one open issue and one question noted by @rpspringuel there that has not (yet) been addressed here.

### CODE REVIEW

**Comment by eroux** [comment]
thanks for cleaning the PR, I wasn't decided between `develop` and `release-4.0`, and made a mistake... I'm glad you're ok for 4.0, as I think it's the last big bug! The people working on a new Antiphonale Monasticum said that apart from this bug, Gregorio is really better than anything else, better than doing things by hand, better than Solesme's books, etc. So I just wanted this bug to be fixed in order for Gregorio to be all that! (and of course, thank you all for making Gregorio such a great software!).

About the duplication noted by @rpspringuel, I'll fix that today, and also document the fact that alterationspace is no more extensible.

**Comment by eroux** [comment]
I think it's ready to be merged now

**Comment by eroux** [comment]
About @rpspringuel 's question: ideally, we could allow `alterationspace` to be a rubber, though it shouldn't change much the look of 99.99% of the scores, but as it's involved in nextbegindifference calculation, I think it would be too risky: having syllables (even slightly) merged would be awful, while having the alterations at a fixed distance will never be... What do you think?

File: CHANGELOG.md [hunk 1]
```diff
@@ -3,7 +3,11 @@ All notable changes to this project will be documented in this file.
 As of v3.0.0 this project adheres to [Semantic Versioning](http://semver.org/). It follows [some conventions](http://keepachangelog.com/).
 
 ## [Unreleased][unreleased]
+### Fixed
+- Spacing was too large when alteration begins a syllable, see [#663](https://github.com/gregorio-project/gregorio/issues/663).
 
+### Changed
+- `alterationspace` is now a fixed dimen, see [UPGRADE.md](UPGRADE.md) for details.
```
Comment (henryso) [inline CHANGELOG.md]: Perhaps "dimension" rather than "dimen"?

**Comment by henryso** [comment]
Aside from my minor comment, this looks fine.  The corresponding tests look good too.

**Comment by eroux** [comment]
thanks, done

**Comment by henryso** [comment]
@rpspringuel If you are OK with the distance not being rubber as explained by @eroux, I think this is good to merge.

**Comment by rpspringuel** [comment]
I'm good.

**Pull request merged**